### PR TITLE
[GEOT-6910] 500+ lines method: Resampler2D:reproject(...)

### PR DESCRIPTION
**Pull request by groldan**
[GEOT-6910](https://osgeo-org.atlassian.net/browse/GEOT-6910) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=geotools&utm_campaign=badge_info)

Refactor `Resampler2D.reproject()` from 500+ lines of code into this:

```
        ResampleState state = resolveParameters(...);

        GridCoverage2D targetCoverage = reuseExistingCoverage(state);
        if (targetCoverage == null) {
            createSourcePlanarImage(state);
            resolveRoiAndNoData(state);

            createTargetToSourceTransform(state);

            computeTargetImageLayout(state);

            targetCoverage = applyJaiOpAndCreateCoverage(state);
        }
        return targetCoverage;
```

<Include a few sentences describing the overall goals for this Pull Request>

# Checklist

Reviewing is a process done by project maintainers, **mostly on a volunteer basis** (thus limited in time). We need to keep the review overhead as small as possible, and appreciate if you help us to do so by completing the following items. Feel free to ask in a comment if you have troubles with any of them.

For all pull requests:

- [ ] Confirm you have read the [contribution guidelines](https://github.com/geotools/geotools/blob/main/CONTRIBUTING.md)
- [ ] You have sent a Contribution Licence Agreement (CLA) as necessary (not required for small changes, e.g., fixing typos in documentation)
- [ ] Make sure the first PR targets the `main` branch, eventual backports will be managed later. This can be ignored if the PR is fixing an issue that only happens in a specific branch, but not in newer ones.
- [ ] The changes are not causing two modules to share the same Java packages (to avoid [Java 9+ split package](http://tutorials.jenkov.com/java/modules.html#split-packages-not-allowed) issues)
- [ ] The changes are not breaking the build in downstream projects using SNAPSHOT dependencies, GeoWebCache and GeoServer (there is an automatic PR check verifying this, check this when it turns green).

The following are required only for core and extension modules (they are welcomed, but not required, for unsupported modules):

- [ ] There is an issue in [Jira](https://osgeo-org.atlassian.net/projects/GEOT) describing the bug/task/new feature (a notable exemptions is, changes not visible to end users). The ticket is for the GeoTools project, if the issue was found elsewhere it's a good practice to link to the origin ticket/issue.
- [ ] The pull request contains changes related to a single objective. If multiple focuses cannot be avoided, each one is in its own commit and has a separate ticket describing it.
- [ ] PR for bug fixes and small new features are presented as a single commit
- [ ] Commit message(s) must be in the form "[GEOT-XYZW] Title of the Jira ticket"
- [ ] New unit tests have been added covering the changes
- [ ] This PR passes all existing unit tests (test results will be reported by Continuous Integration after opening this PR)
- [ ] This PR passes the [QA checks](https://docs.geotools.org/latest/developer/conventions/code/qa.html) (QA checks results will be reported by Continuous Integration after opening this PR)
- [ ] Documentation has been updated accordingly.

Submitting the PR does not require you to check all items, but by the time it gets merged, they should be either satisfied or not applicable.

### CODE REVIEW

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java [hunk 1]
```diff
@@ -243,13 +300,17 @@ public static GridCoverage2D reproject(
...
             final Hints hints,
             final double[] backgroundValues)
             throws FactoryException, TransformException {
+
+        ResampleState state;
+        GridCoverage2D targetCoverage;
+
         ////////////////////////////////////////////////////////////////////////////////////////
         ////                                                                                ////
         //// =======>>  STEP 1: Extracts needed informations from the parameters   <<====== ////
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java:316]:
```suggestion
        ////  STEP 1: Extracts needed information from the parameters  ////
```

This comment has become a summary of what the method does, not indicating the step any more

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java [hunk 2]
```diff
@@ -493,373 +853,235 @@ public static GridCoverage2D reproject(
...
-                    new RenderingHints(
-                            JAI.KEY_INTERPOLATION,
-                            Interpolation.getInstance(Interpolation.INTERP_NEAREST));
-        } else {
-            targetHints.put(
-                    JAI.KEY_INTERPOLATION, Interpolation.getInstance(Interpolation.INTERP_NEAREST));
-        }
-        targetHints.put(JAI.KEY_REPLACE_INDEX_COLOR_MODEL, false);
-        targetHints.put(JAI.KEY_TRANSFORM_ON_COLORMAP, false);
+    private static ResampleState resolveParameters(
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java:908]: This method really seems a constructor for ResampleState.

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java [hunk 3]
```diff
@@ -493,373 +853,235 @@ public static GridCoverage2D reproject(
...
+        @SuppressWarnings("unchecked")
+        Map<String, Object> sourceProps = state.sourceCoverage.getProperties();
+        state.sourceProps = sourceProps != null ? new HashMap<>(sourceProps) : new HashMap<>();
+
+        return state;
+    }
+
+    private static ResampleState resolveRoiAndNoData(ResampleState state) {
+        /*
+         * The projection are usually applied on floating-point values, in order
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java:943]: Hm... this comment seems to be inappropriate here, but it was kind of dangling in the original code too, I cannot find the actual code bit where it applies...

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java [hunk 4]
```diff
@@ -493,373 +853,235 @@ public static GridCoverage2D reproject(
...
 
-        ////////////////////////////////////////////////////////////////////////////////////////
-        ////                                                                                ////
-        ////            STEP 1: Extracts needed informations from the parameters            ////
-        ////            STEP 2: Creates the "target to source" MathTransform                ////
-        //// =======>>  STEP 3: Computes the target image layout                   <<====== ////
-        ////            STEP 4: Applies the JAI operation ("Affine", "Warp", etc)           ////
-        ////                                                                                ////
-        ////////////////////////////////////////////////////////////////////////////////////////
+    private static void createSourcePlanarImage(ResampleState state) {
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java:898]: I don't see expensive code here, could have been part of the state constructor too.

File: modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java [hunk 5]
```diff
@@ -493,373 +853,235 @@ public static GridCoverage2D reproject(
...
+         * NaN values. However, we can apply the projection on integer values if
+         * the interpolation type is "nearest neighbor", since this is not really
+         * an interpolation.
+         *
+         * If this condition is met, then we verify if an "integer version" of the image
+         * is available as a source of the source coverage (i.e. the floating-point image
+         * is derived from the integer image, not the converse).
+         */
+        // Getting optional ROI and NoData as properties
+        Object roiProp = state.sourceProps.get("GC_ROI");
```
Comment (aaime) [inline modules/library/coverage/src/main/java/org/geotools/coverage/processing/operation/Resampler2D.java:954]: This code also seems like it belogs into the state constructor, seems cheap to run.

**Review by aaime — COMMENTED** [review]
Wow this was a hard one to review, diff tools proved basically useless.
Overall I think it looks good, but it also seems unfinished... what you're working towards, here, is a [method object](https://wiki.c2.com/?MethodObject) refactor, you identified a state support, but then all the methods are out of it and reference to its fields, instead of being methods of the state object (which would also simplify all the code, you'd avoid having to refer to the state everywhere).

**Comment by groldan** [comment]
> Wow this was a hard one to review, diff tools proved basically useless.

Imagine what it was to try to understand it to be able of refactoring it :)

I'll either have to wait till having free time to go over your comments, or yield to the original authors for the final touches. Just trying to be proactive in the attempt to activate this PMD rule, instead of saying hey, your method is too long and too difficult to read, please make it better.

**Comment by aaime** [comment]
The original author is long gone, you'll have to push it to over the final line. :-)

**Comment by jodygarnett** [comment]
Do you want to continue with this? Fill in the checklist and we can give it a go..

**Pull request closed without merging**